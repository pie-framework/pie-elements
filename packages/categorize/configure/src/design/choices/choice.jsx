import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import InputHeader from '../input-header';
import { Checkbox } from '@pie-lib/config-ui';
import { DeleteButton } from '../buttons';
import DragHandle from '@material-ui/icons/DragHandle';
import { DragSource } from 'react-dnd';
import debug from 'debug';
import { uid } from '@pie-lib/drag';

const log = debug('@pie-element:categorize:configure:choice');

const canDrag = props => {
  const count = props.choice.categoryCount || 0;
  if (count === 0) {
    return true;
  } else {
    return props.correctResponseCount < count;
  }
};

export class Choice extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    choice: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    connectDragSource: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    correctResponseCount: PropTypes.number.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    })
  };

  static defaultProps = {};

  changeContent = content => {
    const { onChange, choice } = this.props;
    choice.content = content;
    onChange(choice);
  };

  changeCategoryCount = () => {
    const { onChange, choice } = this.props;
    if (choice.categoryCount === 1) {
      choice.categoryCount = 0;
    } else {
      choice.categoryCount = 1;
    }
    onChange(choice);
  };

  render() {
    const {
      classes,
      className,
      choice,
      onDelete,
      connectDragSource,
      connectDragPreview,
      imageSupport
    } = this.props;

    const draggable = canDrag(this.props);

    return (
      <Card className={classNames(classes.choice, className)}>
        <CardActions className={classes.actions}>
          {connectDragSource(
            <span
              className={classNames(
                classes.dragHandle,
                draggable === false && classes.dragDisabled
              )}
            >
              <DragHandle color={draggable ? 'primary' : 'disabled'} />
            </span>
          )}
        </CardActions>
        {connectDragPreview(
          <span>
            <InputHeader
              imageSupport={imageSupport}
              label={choice.content}
              onChange={this.changeContent}
              onDelete={onDelete}
            />
          </span>
        )}

        <CardActions className={classes.actions}>
          <DeleteButton label={'delete'} onClick={onDelete} />
          <Checkbox
            mini
            label={'Remove after placing'}
            checked={choice.categoryCount === 1}
            onChange={this.changeCategoryCount}
          />
        </CardActions>
      </Card>
    );
  }
}
const styles = theme => ({
  actions: {
    padding: 0,
    justifyContent: 'space-between'
  },
  choice: {
    padding: theme.spacing.unit,
    overflow: 'visible'
  },
  dragHandle: {
    cursor: 'move'
  },
  dragDisabled: {
    cursor: 'inherit'
  }
});

const StyledChoice = withStyles(styles)(Choice);

export const spec = {
  canDrag,
  beginDrag: props => {
    const out = {
      id: props.choice.id
    };
    log('[beginDrag] out:', out);
    return out;
  },
  endDrag: (props, monitor) => {
    if (!monitor.didDrop()) {
      const item = monitor.getItem();
      if (item.categoryId) {
        log('wasnt droppped - what to do?');
        props.onRemoveChoice(item);
      }
    }
  }
};

const DraggableChoice = DragSource(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
    isDragging: monitor.isDragging()
  })
)(StyledChoice);

export default uid.withUid(DraggableChoice);
