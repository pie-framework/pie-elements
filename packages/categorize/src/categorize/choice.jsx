import React from 'react';
import PropTypes from 'prop-types';
import { useDrag } from 'react-dnd';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import { color } from '@pie-lib/pie-toolbox/render-ui';

export const ChoiceType = {
  content: PropTypes.string.isRequired,
  id: PropTypes.string,
};

export const Layout = ({ classes, className, content, isDragging, disabled, correct }) => {
  const rootNames = classNames(
      correct === true && 'correct',
      correct === false && 'incorrect',
      classes.choice,
      isDragging && classes.dragging,
      disabled && classes.disabled,
      className,
  );
  const cardNames = classNames(classes.card);
  return (
      <div className={rootNames}>
        <Card className={cardNames}>
          <CardContent
              classes={{ root: classes.cardRoot }}
              dangerouslySetInnerHTML={{ __html: content }}
          />
        </Card>
      </div>
  );
};

Layout.propTypes = {
  ...ChoiceType,
  classes: PropTypes.object.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  correct: PropTypes.bool,
};

const styles = (theme) => ({
  choice: {
    direction: 'initial',
    cursor: 'pointer',
    width: '100%',
    '&.correct': {
      border: `solid 2px ${color.correct()}`,
    },
    '&.incorrect': {
      border: `solid 2px ${color.incorrect()}`,
    },
    borderRadius: '6px',
  },
  cardRoot: {
    color: color.text(),
    backgroundColor: color.white(),
    fontSize: theme.typography.fontSize + 2,
    '&:last-child': {
      paddingBottom: theme.spacing.unit * 2,
    },
    borderRadius: '4px',
    border: '1px solid',
  },
  disabled: {
    cursor: 'not-allowed',
    opacity: '0.6',
  },
  dragging: {
    cursor: 'move',
  },
  card: {
    color: color.text(),
    backgroundColor: color.background(),
    width: '100%',
    pointerEvents: 'none',
  },
});

const Styled = withStyles(styles)(Layout);

const Choice = ({ id, content, disabled, correct, extraStyle, onRemoveChoice }) => {
  const [{ isDragging }, drag] = useDrag({
    type: 'categorize',
    item: { id, content },
    canDrag: !disabled,
    end: (item, monitor) => {
      if (!monitor.didDrop()) {
        onRemoveChoice(item);
      }
    },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  return (
      <div ref={drag} style={{ margin: '4px', ...extraStyle }}>
        <Styled
            id={id}
            content={content}
            disabled={disabled}
            correct={correct}
            isDragging={isDragging}
        />
      </div>
  );
};

Choice.propTypes = {
  ...ChoiceType,
  extraStyle: PropTypes.object,
  onRemoveChoice: PropTypes.func.isRequired,
};

export default Choice;
