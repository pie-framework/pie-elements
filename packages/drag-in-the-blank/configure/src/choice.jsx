import React from 'react';
import MoreVert from '@material-ui/icons/MoreVert';
import Delete from '@material-ui/icons/Delete';
import { DragSource } from '@pie-lib/drag';
import { withStyles } from '@material-ui/core/styles';
import { choiceIsEmpty } from './markupUtils';

const GripIcon = ({ style }) => {
  return (
    <span style={style}>
      <MoreVert
        style={{
          margin: '0 -16px'
        }}
      />
      <MoreVert />
    </span>
  );
};

export const BlankContent = withStyles(theme => ({
  choice: {
    border: `solid 0px ${theme.palette.primary.main}`
  },
  disabled: {}
}))(props => {
  const { connectDragSource, choice, onClick, onRemoveChoice } = props;

  return connectDragSource(
    <div
      style={{
        display: 'inline-flex',
        minWidth: '178px',
        minHeight: '36px',
        background: '#FFF',
        border: '1px solid #C0C3CF',
        boxSizing: 'border-box',
        borderRadius: '3px',
        overflow: 'hidden',
        position: 'relative',
        padding: '8px 35px 8px 35px'
      }}
      onClick={onClick}
    >
      <GripIcon
        style={{
          position: 'absolute',
          top: '6px',
          left: '15px',
          color: '#9B9B9B',
          zIndex: 2
        }}
      />
      <span
        dangerouslySetInnerHTML={{
          __html: choice.value
        }}
      />
      <Delete
        style={{
          position: 'absolute',
          top: '6px',
          right: '0',
          color: '#9B9B9B',
          zIndex: 2
        }}
        onClick={(e) => {
          e.preventDefault();
          e.stopPropagation();

          onRemoveChoice(e);
        }}
      />
    </div>
  );
});

const tileSource = {
  canDrag(props) {
    if (choiceIsEmpty(props.choice)) {
      alert('You need to define a value for an answer choice before it can be associated with a response area.');
      return false;
    }

    return !props.disabled;
  },
  beginDrag(props) {
    return {
      id: props.targetId,
      value: props.choice,
      instanceId: props.instanceId
    };
  }
};

export default DragSource('drag-in-the-blank-choice', tileSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
}))(BlankContent);
