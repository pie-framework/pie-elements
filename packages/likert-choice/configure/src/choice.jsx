import React, {Component, Fragment} from 'react';
import {NChoice} from '@pie-lib/config-ui';

export default class Choice extends Component {

    constructor(props) {
        super(props);
        this.state = {response: '', choice: '', type: ''}
    }

    render() {
        return (
            <Fragment>
                <NChoice
                    header="Response Type"
                    value={this.state.response}
                    onChange={response => this.setState({response})}
                    opts={[
                        {label: 'Likert 3', value: '3'},
                        {label: 'Likert 5', value: '5'},
                        {label: 'Likert 7', value: '7'}
                    ]}
                />
                <NChoice
                    header="Choice Labels"
                    value={this.state.choice}
                    onChange={choice => this.setState({choice})}
                    opts={[
                        {label: 'Numbers', value: 'numbers'},
                        {label: 'Letters', value: 'letters'},
                        {label: 'Graphics', value: 'graphics'},
                        {label: 'None', value: 'none'}
                    ]}
                />
                <NChoice
                    header="Label Type"
                    value={this.state.type}
                    onChange={type => this.setState({type})}
                    opts={[
                        {label: 'Agreement', value: 'agreement'},
                        {label: 'Frequency', value: 'frequency'},
                        {label: 'Yes/No', value: 'yesno'},
                        {label: 'Likelihood', value: 'likelihood'},
                        {label: 'Importance', value: 'importance'},
                        {label: 'None', value: 'none'},
                        {label: 'Custom', value: 'custom'}
                    ]}
                />
            </Fragment>
        )
    }

}