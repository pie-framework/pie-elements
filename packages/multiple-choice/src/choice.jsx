import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import ChoiceInput from './choice-input';

export class Choice extends React.Component {
  onChange = (choice) => {
    const { disabled, onChoiceChanged } = this.props;

    if (!disabled) {
      onChoiceChanged(choice);
    }
  };

  render() {
    const {
      choice,
      index,
      choicesLength,
      showCorrect,
      isEvaluateMode,
      choiceMode,
      disabled,
      checked,
      correctness,
      displayKey,
      classes,
      choicesLayout,
      gridColumns,
      updateSession,
    } = this.props;
    const choiceClass = 'choice' + (index === choicesLength - 1 ? ' last' : '');

    const feedback = !isEvaluateMode || showCorrect ? '' : choice.feedback;

    const choiceProps = {
      ...choice,
      checked,
      choiceMode,
      disabled,
      feedback,
      correctness,
      displayKey,
      choicesLayout,
      gridColumns,
      updateSession,
      onChange: this.onChange,
      isEvaluateMode,
    };

    const names = classNames(classes.choice, {
      [classes.noBorder]: index === choicesLength - 1 || choicesLayout !== 'vertical',
      [classes.horizontalLayout]: choicesLayout === 'horizontal',
    });

    return (
      <div className={choiceClass} key={index}>
        <ChoiceInput {...choiceProps} className={names} />
      </div>
    );
  }
}

Choice.propTypes = {
  choiceMode: PropTypes.oneOf(['radio', 'checkbox']),
  choice: PropTypes.object,
  disabled: PropTypes.bool.isRequired,
  onChoiceChanged: PropTypes.func,
  classes: PropTypes.object.isRequired,
  index: PropTypes.number,
  choicesLength: PropTypes.number,
  showCorrect: PropTypes.bool,
  isEvaluateMode: PropTypes.bool,
  checked: PropTypes.bool,
  correctness: PropTypes.string,
  displayKey: PropTypes.string,
  choicesLayout: PropTypes.oneOf(['vertical', 'grid', 'horizontal']),
  gridColumns: PropTypes.string,
};

export default withStyles((theme) => ({
  choice: {
    paddingTop: theme.spacing.unit * 2.5,
    paddingBottom: theme.spacing.unit + 2,
    borderBottom: `1px solid ${theme.palette.grey[300]}`,
  },
  noBorder: {
    borderBottom: 'none',
  },
  horizontalLayout: {
    paddingRight: theme.spacing.unit * 2.5,
    '& label': {
      marginRight: theme.spacing.unit,
      // '& span:first-child': {
      //   paddingRight: 0
      // }
    },
  },
}))(Choice);
