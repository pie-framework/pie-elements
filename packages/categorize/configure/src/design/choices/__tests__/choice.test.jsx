import { shallow } from 'enzyme';
import React from 'react';
import { Choice, spec } from '../choice';

describe('choice', () => {
  let onChange;
  let onDelete;
  let connectDragSource;
  let connectDragPreview;

  beforeEach(() => {
    onChange = jest.fn();
    onDelete = jest.fn();
    connectDragSource = jest.fn();
    connectDragPreview = jest.fn();
  });

  const wrapper = (extras) => {
    const props = {
      classes: {},
      correctResponseCount: 0,
      className: 'classname',
      choice: {
        content: 'hi',
        id: '1',
      },
      onChange,
      onDelete,
      connectDragSource,
      connectDragPreview,
      ...extras,
    };

    return shallow(<Choice {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      const w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('changeContent', () => {
      it('calls onChange', () => {
        const w = wrapper();
        w.instance().changeContent('foo');
        expect(onChange).toBeCalledWith({ content: 'foo', id: '1' });
      });
    });
    describe('changeCategoryCount', () => {
      it('calls onChange', () => {
        const w = wrapper();
        w.instance().changeCategoryCount(1);
        expect(onChange).toBeCalledWith({
          content: 'hi',
          id: '1',
          categoryCount: 1,
        });
      });
    });
  });
});

describe('spec', () => {
  describe('canDrag', () => {
    it('returns true for categoryCount 0 and correctResponseCount: 100', () => {
      const result = spec.canDrag({
        choice: { categoryCount: 0 },
        correctResponseCount: 100,
      });
      expect(result).toEqual(true);
    });

    it('returns true for categoryCount 1 and correctResponseCount: 0', () => {
      const result = spec.canDrag({
        choice: { categoryCount: 1 },
        correctResponseCount: 0,
      });
      expect(result).toEqual(true);
    });

    it('returns false for categoryCount 1 and correctResponseCount: 1', () => {
      const result = spec.canDrag({
        choice: { categoryCount: 1 },
        correctResponseCount: 1,
      });
      expect(result).toEqual(false);
    });
  });

  describe('beginDrag', () => {
    it('returns the id', () => {
      const result = spec.beginDrag({ choice: { id: '1' } });
      expect(result).toEqual({ id: '1' });
    });
  });

  describe('endDrag', () => {
    let monitor;
    let item;
    beforeEach(() => {
      item = {
        categoryId: '2',
      };
      monitor = {
        didDrop: jest.fn().mockReturnValue(false),
        getItem: jest.fn().mockReturnValue(item),
      };
    });

    it('calls onRemoveChoice if !didDrop is false and categoryId is defined', () => {
      const props = {
        onRemoveChoice: jest.fn(),
      };
      spec.endDrag(props, monitor);
      expect(props.onRemoveChoice).toBeCalledWith(item);
    });
  });
});
