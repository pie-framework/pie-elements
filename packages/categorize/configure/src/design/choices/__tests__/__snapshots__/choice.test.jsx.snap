// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choice snapshot renders 1`] = `
<WithStyles(Card)
  className="classname"
>
  <WithStyles(CardActions) />
  <WithStyles(CardActions)>
    <WithStyles(Component)
      label="delete"
      onClick={[MockFunction]}
    />
  </WithStyles(CardActions)>
</WithStyles(Card)>
`;
